### PR TITLE
Wg 43a add workout to favourites

### PR DESCRIPTION
**Pull request by chrismcanulty**
PR where one workout only can be added to favourites. Next PR will focus on saving multiple workouts to favourites via Async Storage (up to a limit of 10)
All CRUD operations are possible within the favourite workout itself and it renders the same way as a newly generated workout

JIRA ticket: https://lhl-midterm-ecommerce.atlassian.net/browse/WG-43

Screenshots

Screen Shot 2023-05-22 at 8 28 20 PM

Screen Shot 2023-05-22 at 8 28 49 PM

### CODE REVIEW

File: src/component/ExerciseData.tsx [hunk 1]
```diff
@@ -9,6 +9,8 @@ export default function ExerciseData({
   item: any;
   workoutId: String;
 }) {
+  console.log('item is?', item);
```
Comment (wadesun18) [inline src/component/ExerciseData.tsx]: Remove

File: src/component/FavouriteExerciseData.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import {ExerciseSet} from '../../types/data';
+
+export default function FavouriteExerciseData({
+  item,
+  workoutId,
+}: {
+  item: any;
+  workoutId: String;
+}) {
+  console.log('item is?', item);
```
Comment (wadesun18) [inline src/component/FavouriteExerciseData.tsx]: Remove

File: src/component/FavouriteWorkout.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      let values = await AsyncStorage.getItem('@workout_key');
+      if (values !== null) {
+        const favourites = JSON.parse(values);
+        setFavouriteWorkoutData(favourites);
+      }
+    } catch (e) {
+      // read error
+    }
+
+    console.log('Done.');
```
Comment (wadesun18) [inline src/component/FavouriteWorkout.tsx]: Remove

File: src/component/FavouriteWorkout.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    getFavouriteWorkoutData();
+  }, []);
+
+  if (favouriteExerciseData.length === 0) {
+    return null;
+  }
+
+  if (favouriteWorkoutData.length === 0) {
+    return null;
+  }
```
Comment (wadesun18) [inline src/component/FavouriteWorkout.tsx:98]: Combine both

File: src/component/WorkoutExercise.tsx [hunk 5]
```diff
@@ -87,6 +87,8 @@ export default function WorkoutExercise({
 
   const headers = ['Set', 'Reps', 'Weight', 'Done', 'Edit', 'Delete'];
 
+  console.log('workout', workout);
```
Comment (wadesun18) [inline src/component/WorkoutExercise.tsx]: Remove

File: src/screens/SavedWorkouts.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    getFavouriteWorkoutData();
+  }, []);
+
+  if (favouriteExerciseData.length === 0) {
+    return null;
+  }
+
+  if (favouriteWorkoutData.length === 0) {
+    return null;
+  }
```
Comment (wadesun18) [inline src/screens/SavedWorkouts.tsx:98]: Combine

**Review by wadesun18 — CHANGES_REQUESTED** [review]

File: src/component/FavouriteExerciseData.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+}: {
+  item: any;
+  workoutId: String;
+}) {
+  if (!item) return null;
+
+  return item.map((row: ExerciseSet, index: Number) => {
+    if (!row) {
+      return null;
+    }
```
Comment (wadesun18) [inline src/component/FavouriteExerciseData.tsx]: This check should be inside FavouriteTableRow

File: src/component/FavouriteTableRow.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  workoutId: String;
+}) {
+  const {favouriteClickComplete, favouriteDeleteSet, favouriteEditSet} =
+    useContext(UserContext);
+
+  const [reps, setReps] = useState('10');
+  const [weight, setWeight] = useState('0');
+
+  const onChangedReps = (text: string) => {
+    isNaN(Number(text)) || text.includes('.')
```
Comment (wadesun18) [inline src/component/FavouriteTableRow.tsx:49]: It's unclear what you are trying to do here

File: src/component/FavouriteTableRow.tsx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+      ? (text = text.substr(0, text.length - 1))
+      : text;
+
+    setReps(text);
+  };
+
+  const onChangedWeight = (text: string) => {
+    isNaN(Number(text)) ? (text = text.substr(0, text.length - 1)) : text;
```
Comment (wadesun18) [inline src/component/FavouriteTableRow.tsx:59]: same

File: src/component/MultipleSelectList.tsx [hunk 10]
```diff
@@ -415,8 +415,6 @@ export const MultipleSelectList: React.FC<MultipleSelectListProps> = ({
                         onPress={() => {
                           let existing = selectedval?.indexOf(value);
 
-                          // console.log(existing);
```
Comment (wadesun18) [inline src/component/MultipleSelectList.tsx:418]: remove

File: src/context/User.Context.js [hunk 11]
```diff
@@ -201,6 +319,23 @@ const UserProvider = ({children}) => {
...
+          favouriteClickComplete,
+          favouriteDeleteSet,
+          favouriteEditSet,
+          getFavouriteTokens,
+          favouriteTokens,
+          title,
+          setTitle,
+          workoutNames,
+          setWorkoutNames,
+          getWorkoutNames,
```
Comment (wadesun18) [inline src/context/User.Context.js:338]: I think your user context is getting really large. It's a good idea to separate it out like FavouriteContext

File: src/screens/SavedWorkouts.tsx [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    getFavouriteWorkoutData();
+  }, []);
+
+  if (favouriteExerciseData.length === 0 || favouriteWorkoutData.length === 0) {
+    return null;
+  }
+
+  return (
+    <ContainerWrapper>
+      {/* <Text>{title}</Text> */}
```
Comment (wadesun18) [inline src/screens/SavedWorkouts.tsx]: remove

**Review by wadesun18 — APPROVED** [review]

File: src/component/FavouriteTableRow.tsx (hunk 8 above)
Comment (chrismcanulty) [reply src/component/FavouriteTableRow.tsx:49]: Prevent user from being able to input partial reps or non-integer reps. Added comment above the const, as well as onChangedWeight which is similar

File: src/context/User.Context.js (hunk 11 above)
Comment (chrismcanulty) [reply src/context/User.Context.js:338]: Noted for future projects, should be very little to no additional state or functions needed for this project

**Pull request merged**